pkg/storage/formattedstore/formattedstore.go

```diff
@@ -0,0 +1,178 @@
+	store, err := p.provider.OpenStore(name)
+	if err != nil {
+		return nil, fmt.Errorf(failOpenUnderlyingStore, err)
+	}
+
+	formatStore := formatStore{
+		store:                 store,
+		formatter:             p.formatter,
+		skipIteratorFiltering: p.skipIteratorFiltering,
+	}
+
+	return &formatStore, nil
+}
+
+// CloseStore closes the store with the given name in the underlying provider.
+func (p *FormattedProvider) CloseStore(name string) error {
+	err := p.provider.CloseStore(name)
+	if err != nil {
+		return fmt.Errorf(failCloseUnderlyingStore, err)
+	}
+
+	return p.provider.CloseStore(name)
+}
+
+// Close closes all stores created in the underlying provider.
+func (p *FormattedProvider) Close() error {
+	err := p.provider.Close()
+	if err != nil {
+		return fmt.Errorf(failCloseAllUnderlyingStores, err)
+	}
+
+	return p.provider.Close()
+}
+
+type formatStore struct {
+	store                 storage.Store
+	formatter             Formatter
+	skipIteratorFiltering bool
+}
+
+func (s *formatStore) Put(k string, v []byte) error {
+	formattedValue, err := s.formatter.FormatPair(k, v)
+	if err != nil {
+		return fmt.Errorf(failFormat, err)
+	}
+
+	err = s.store.Put(k, formattedValue)
+	if err != nil {
+		return fmt.Errorf(failPutInUnderlyingStore, err)
+	}
+
+	return nil
+}
+
+func (s *formatStore) Get(k string) ([]byte, error) {
+	formattedValue, err := s.store.Get(k)
+	if err != nil {
+		return nil, fmt.Errorf(failGetFromUnderlyingStore, err)
+	}
+
+	_, value, err := s.formatter.ParsePair(formattedValue)
+	if err != nil {
+		return nil, fmt.Errorf(failParseFormattedValue, err)
+	}
+
+	return value, nil
+}
+
+// The "don't filter" switch is here since it's possible for the underlying storage to have already done
+// the filtering, in which case it would be redundant to do it here too.
+// TODO (#2315) While doing the storage interface rework, see if the skipIteratorFiltering switch is still
+//  needed.
+func (s *formatStore) Iterator(startKey, endKey string) storage.StoreIterator {
+	formattedIterator := s.store.Iterator(startKey, endKey)
+
+	err := formattedIterator.Error()
+	if err != nil {
+		return mem.NewMemIterator(nil, fmt.Errorf(failGetIteratorFromUnderlyingStore, err))
+	}
+
+	var batch [][]string
+
+	for formattedIterator.Next() {
+		key, value, err := s.formatter.ParsePair(formattedIterator.Value())
+		if err != nil {
+			return mem.NewMemIterator(nil, fmt.Errorf(failParseFormattedValue, err))
+		}
+
+		if s.skipIteratorFiltering {
+			batch = append(batch, []string{key, string(value)})
+		} else {
+			if endKey == "" {
+				return mem.NewMemIterator(nil, nil)
+			}
+
+			if strings.HasPrefix(key, strings.TrimSuffix(endKey, storage.EndKeySuffix)) && key != endKey {
+				batch = append(batch, []string{key, string(value)})
+				continue
+			}
+
+			if key >= startKey && key < endKey {
+				batch = append(batch, []string{key, string(value)})
+			}
+		}
+	}
+
+	formattedIterator.Release()
```

Comment:
I would appreciate an extra set of eyes on this filter logic. My goal was to ensure that the filtering logic was consistent with the mem store version, however the major difference here is that the way I've implemented it doesn't require sorting all the keys. I think the logic is kind of messy, though. Is there a better way to do it?

@fqutishat, @soluchok , I saw you were both involved in Iterator method work in the past, can you take a look?

Also note that I added a bunch of tests for this, but I used the same test style as the existing mem store test so I could compare my results to ensure consistency.